rclpy/src/rclpy/action_client.cpp

```diff
@@ -0,0 +1,335 @@
+    return py::make_tuple(py::none(), py::none()); \
+  } else if (RCL_RET_OK != ret) { \
+    throw rclpy::RCLError("Failed to take " #Type); \
+  } \
+  return py::make_tuple(py::int_(sequence), convert_to_py(taken_msg.get(), pymsg_type)); \
+
+py::tuple
+ActionClient::take_goal_response(py::object pymsg_type)
+{
+  TAKE_SERVICE_RESPONSE(goal)
+}
+
+#define SEND_SERVICE_REQUEST(Type) \
+  destroy_ros_message_signature * destroy_ros_message = NULL; \
+  void * raw_ros_request = rclpy_convert_from_py(pyrequest.ptr(), &destroy_ros_message); \
+  if (!raw_ros_request) { \
+    throw py::error_already_set(); \
+  } \
+  int64_t sequence_number; \
+  rcl_ret_t ret = rcl_action_send_ ## Type ## _request( \
+    rcl_action_client_.get(), raw_ros_request, &sequence_number); \
+  destroy_ros_message(raw_ros_request); \
+  if (RCL_RET_OK != ret) { \
+    throw rclpy::RCLError("Failed to send " #Type " request"); \
+  } \
+  return sequence_number;
+
+int64_t
+ActionClient::send_result_request(py::object pyrequest)
+{
+  SEND_SERVICE_REQUEST(result);
+}
+
+py::tuple
+ActionClient::take_cancel_response(py::object pymsg_type)
+{
+  TAKE_SERVICE_RESPONSE(cancel)
+}
+
+#define TAKE_MESSAGE(Type) \
+  auto taken_msg = create_from_py(pymsg_type); \
+  rcl_ret_t ret = rcl_action_take_ ## Type(rcl_action_client_.get(), taken_msg.get()); \
+  if (RCL_RET_OK != ret) { \
+    if (RCL_RET_ACTION_CLIENT_TAKE_FAILED == ret) { \
+      /* if take failed, just do nothing */ \
+      return py::none(); \
+    } \
+    throw rclpy::RCLError("Failed to take " #Type " with an action client"); \
+  } \
+  return convert_to_py(taken_msg.get(), pymsg_type);
+
+py::object
+ActionClient::take_feedback(py::object pymsg_type)
+{
+  TAKE_MESSAGE(feedback)
+}
+
+py::object
+ActionClient::take_status(py::object pymsg_type)
+{
+  TAKE_MESSAGE(status)
+}
+
+int64_t
+ActionClient::send_cancel_request(py::object pyrequest)
+{
+  SEND_SERVICE_REQUEST(cancel)
+}
+
+int64_t
+ActionClient::send_goal_request(py::object pyrequest)
+{
+  SEND_SERVICE_REQUEST(goal)
+}
+
+py::tuple
+ActionClient::take_result_response(py::object pymsg_type)
+{
+  TAKE_SERVICE_RESPONSE(result);
+}
+
+py::tuple
+ActionClient::get_num_entities()
+{
+  size_t num_subscriptions = 0u;
+  size_t num_guard_conditions = 0u;
+  size_t num_timers = 0u;
+  size_t num_clients = 0u;
+  size_t num_services = 0u;
+
+  rcl_ret_t ret;
+  ret = rcl_action_client_wait_set_get_num_entities(
+    rcl_action_client_.get(),
+    &num_subscriptions,
+    &num_guard_conditions,
+    &num_timers,
+    &num_clients,
+    &num_services);
+  if (RCL_RET_OK != ret) {
+    std::string error_text{"Failed to get number of entities for 'rcl_action_client_t'"};
+    throw rclpy::RCLError(error_text);
+  }
+
+  return py::make_tuple(
+    py::int_(num_subscriptions), py::int_(num_guard_conditions), py::int_(num_timers),
+    py::int_(num_clients), py::int_(num_services));
```

Comment:
@ahcorde I'd expect:
```suggestion
  return py::make_tuple(
    num_subscriptions, num_guard_conditions, num_timers,
    num_clients, num_services);
```
to work.